File: src/main/java/com/dyrosoft/kvbparser/parser/StationsParserFuc.java
package com.dyrosoft.kvbparser.parser;

import com.dyrosoft.kvbparser.models.Station;
import com.google.common.base.Strings;
import com.google.common.collect.ImmutableList;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class StationsParserFuc extends AbstractRxHtmlParserFunc<ImmutableList<Station>> {

    @Override
    protected ImmutableList<Station> parse(final String html) {
        final ImmutableList.Builder<Station> builder = ImmutableList.builder();

        final Document document = Jsoup.parse(html);
        final Elements elements = document.getElementsByTag("a");
        for (final Element element : elements) {
            final String href = element.attr("href");
            if (Strings.isNullOrEmpty(href)) {
                continue;
            }

            final Pattern pattern = Pattern.compile("/german/hst/overview/(\\d+)/");
            final Matcher matcher = pattern.matcher(href);
            if (!matcher.matches()) {
                continue;
            }

            final String id =
                    href.replace("german/hst/overview/", "").replace("/", "");

            builder.add(new Station(id, element.text()));
        }

        return builder.build();
    }
}

File: src/main/java/com/dyrosoft/kvbparser/parser/StationsParserFuc.kt
package com.dyrosoft.kvbparser.parser

import com.dyrosoft.kvbparser.models.Station
import org.jsoup.Jsoup
import java.util.regex.Pattern

class StationsParserFuc : AbstractRxHtmlParserFunc<List<Station>>() {

    override fun parse(html: String): List<Station> {
        val list = mutableListOf<Station>()
        val document = Jsoup.parse(html)
        val elements = document.getElementsByTag("a")
        for (element in elements) {
            val href = element.attr("href")
            if (href.isNullOrEmpty()) {
                continue
            }

            val pattern = Pattern.compile("/haltestellen/overview/(\\d+)/")
            val matcher = pattern.matcher(href)
            if (!matcher.matches()) {
                continue
            }

            val id = href.replace("haltestellen/overview/", "").replace("/", "")

            list.add(Station(id, element.text()))
        }

        return list.toList()
    }
}
